File: chapter02/problem_h/solution.cpp
#include<cstdio>
#include<iostream>

using namespace std;

/**
* 왼쪽 아래 좌표가 (x,y)인 픽셀이 반지름 R인 원에 포함되는가?
* @param x
* @param y
* @param R
* @return  포함된다면 true, else false
*/
bool isInside(long long x, long long y, long long R) {
	long long sqd = x * x + y * y; //거리의 제곱
	if (sqd < R * R)  //반지름의 제곱
	{   //원점과의 거리가 반지름보다 작다면, 원 안에 있다.
		return true;
	}
	return false;
}

void testcase(int caseIndex) {
	long long R;
	scanf("%lld", &R);

	long long sum = 0; //1사분면에 존재하는 총 픽셀의 수
	long long y = R;
	for (long x = 0; x <= R; x++) {
		long long height = 0;
		for (; y >= 0; y--) {
			if (isInside(x, y, R)) {   //위에서 부터 내려오다가
				//가장 최초로 원 안에 포함된 픽셀 (x, y)
				//이 그룹의 높이는 (y+1)이 된다.
				height = (y + 1);
				break;
			}
		}

		sum += height; //너비는 1이므로
	}


	printf("#%d\n", caseIndex);
	printf("%lld\n", sum * 4); //모든 사분면의 픽셀 수
}

int main() {
	int caseSize;
	scanf("%d", &caseSize);
	for (int caseIndex = 1; caseIndex <= caseSize; caseIndex += 1) {
		testcase(caseIndex);
	}
	return 0;
}

File: chapter02/problem_h/solution.py
def is_inside(x: int, y: int, R: int) -> bool:
	"""
	픽셀 (x, y)가 반지름이 R인 원에 포함되는지 여부를 반환하는 함수

	:param x: 	정수 x 좌표
	:param y: 	정수 y 좌표
	:param R: 	정수 반지름
	:return: 	픽셀이 원에 포함되면 True, 아니면 False
	"""
	return x ** 2 + y ** 2 < R ** 2


def get_number_of_pixels(R: int) -> int:
	"""
	반지름이 R인 원에 포함되는 픽셀의 수를 반환하는 함수

	:param R: 	정수 반지름
	:return: 	픽셀의 수
	"""

	pixels = 0
	h = R + 1  # 하강 검사를 시작할 높이
	for x in range(0, R + 1):
		for y in range(h, -1, -1):
			if is_inside(x, y, R):  # (x, y)가 최초로 원에 포함되면
				pixels += y + 1  # 히스토그램의 높이를 더한다
				h = y
				break

	return pixels * 4


if __name__ == "__main__":
	case_size = int(input())
	for case_index in range(1, case_size + 1):
		radius = int(input())
		print(f"#{case_index}")
		print(f"{get_number_of_pixels(radius)}")
